SearchActivity.py

```diff
@@ -185,144 +254,61 @@ def _write_scores_to_clipboard(self, button=None):
-            id = self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].OfferDBusTube(
-                SERVICE, {})
-        else:
-            _logger.debug('I am joining an activity: waiting for a tube...')
-            self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].ListTubes(
-                reply_handler=self._list_tubes_reply_cb,
-                error_handler=self._list_tubes_error_cb)
-        self._game.set_sharing(True)
-
-    def _list_tubes_reply_cb(self, tubes):
-        """ Reply to a list request. """
-        for tube_info in tubes:
-            self._new_tube_cb(*tube_info)
-
-    def _list_tubes_error_cb(self, e):
-        """ Log errors. """
-        _logger.debug('Error: ListTubes() failed: %s' % (e))
-
-    def _new_tube_cb(self, id, initiator, type, service, params, state):
-        """ Create a new tube. """
-        _logger.debug('New tube: ID=%d initator=%d type=%d service=%s \
-params=%r state=%d' % (id, initiator, type, service, params, state))
-
-        if (type == TelepathyGLib.TubeType.DBUS and service == SERVICE):
-            if state == TelepathyGLib.TubeState.LOCAL_PENDING:
-                self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].AcceptDBusTube(id)
-
-            tube_conn = TubeConnection(
-                self.conn, self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES], id,
-                group_iface=self.text_chan[
-                    TelepathyGLib.IFACE_CHANNEL_INTERFACE_GROUP])
-
-            self.chattube = ChatTube(tube_conn, self.initiating,
-                                     self.event_received_cb)
+        _logger.debug('received_restore %s' % (payload))
+        self._game.restore_game(json_load(payload))
 
     def _setup_dispatch_table(self):
         ''' Associate tokens with commands. '''
         self._processing_methods = {
             'n': [self._receive_new_game, 'get a new game grid'],
             'p': [self._receive_dot_click, 'get a dot click'],
+            'N': [self._receive_nick, 'receive nick from opponent'],
+            'j': [self._receive_join, 'receive new joiner'],
+            'r': [self._receive_restore, 'restore game state'],
         }
 
-    def event_received_cb(self, event_message):
-        ''' Data from a tube has arrived. '''
-        if len(event_message) == 0:
-            return
-        try:
-            command, payload = event_message.split('|', 2)
-        except ValueError:
-            _logger.debug('Could not split event message %s' % (event_message))
-            return
-        self._processing_methods[command][0](payload)
+    def send_event(self, command, payload):
+        ''' Send event through the tube. '''
+        if hasattr(self, 'collab') and self.collab is not None:
+            self.collab.post(dict(
+                command=command,
+                payload=payload
+            ))
 
     def send_new_game(self):
```

Comment:
You can also remove this function and `post` the needed message whenever this was called.